runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -0,0 +1,486 @@
+    if events.rename then
+      local _, staterr, staterrname = uv.fs_stat(fullpath)
+      if staterrname == 'ENOENT' then
+        change_type = protocol.FileChangeType.Deleted
+      else
+        assert(not staterr, staterr)
+        change_type = protocol.FileChangeType.Created
+      end
+    end
+
+    for _, regs in pairs(watched_paths[path].callbacks) do
+      for _, reg in pairs(regs) do
+        reg.callback(fullpath, change_type)
+      end
+    end
+
+    local stat, staterr, staterrname = uv.fs_stat(path)
+    if staterrname == 'ENOENT' then
+      local _, stoperr = fsevent:stop()
+      assert(not stoperr, stoperr)
+      fsevent:close()
+      watched_paths[path] = nil
+      return
+    end
+    assert(not staterr, staterr)
+    if stat.ino ~= watched_paths[path].inode then
+      watched_paths[path].inode = stat.ino
+      local _, stoperr = fsevent:stop()
+      assert(not stoperr, stoperr)
+      fsevent:close()
+      watched_paths[path].fsevent = start_watch(path)
+    end
+
+    if not recursive_watch and stat.type == 'directory' then
+      for client_id, regs in pairs(registrations) do
+        for reg_id, watchers in pairs(regs) do
+          for _, watcher in ipairs(watchers) do
+            if vim.startswith(path .. path_sep, watcher.base_dir .. path_sep) then
+              fsevent_ensure_recursive(
+                path,
+                watcher.pattern,
+                watcher.kind,
+                client_id,
+                reg_id,
+                stat.type
+              )
+            end
+          end
+        end
+      end
+    end
+  end)
+  return fsevent
+end
+
+---@private
+--- Creates libuv fs_events handles.
+---
+---@param path string Absolute path to watch.
+---@param pattern string|table The LSP glob pattern (raw or parsed) to match against.
+---@param kind number The LSP WatchKind value.
+---@param client_id number The LSP client's ID.
+---@param reg_id string The ID used to register the request.
+function fsevent_ensure_recursive(path, pattern, kind, client_id, reg_id, filetype)
+  if filetype == 'directory' or not watch_each_file or M._match(pattern, path) then
+    local new_path = false
+    if not watched_paths[path] then
+      new_path = true
+
+      local stat, fserr = uv.fs_stat(path)
+      assert(not fserr, fserr)
+      watched_paths[path] = {
+        fsevent = start_watch(path),
+        inode = stat.ino,
+        callbacks = {},
+      }
+    end
+
+    watched_paths[path].callbacks[client_id] = watched_paths[path].callbacks[client_id] or {}
+    watched_paths[path].callbacks[client_id][reg_id] = watched_paths[path].callbacks[client_id][reg_id]
+      or {}
+    watched_paths[path].callbacks[client_id][reg_id].callback = watched_paths[path].callbacks[client_id][reg_id].callback
+      or get_callback(path, client_id, reg_id)
+    watched_paths[path].callbacks[client_id][reg_id].filters = watched_paths[path].callbacks[client_id][reg_id].filters
+      or {}
```

Comment:
Could avoid the redundant table lookups with something like this:

```suggestion
    local watched_path = watched_paths[path]
    local callback = watched_path.callbacks[client_id] or {}
    watched_path.callbacks[client_id] = callback
    local registration = callback[reg_id] or {}
    callback[reg_id] = registration
    registration.callback = registration.callback or get_callback(path, client_id, reg_id)
    registration.filters = registration.filters or {}
```

And also use the local variables further below

Reply:
That's definitely better, I'll do that.